File: JavaScript/SignalGhost.js
import { Ghost } from "./Ghost"

export class SignalGhost extends Ghost
{
	constructor()
	{
		this.color = "#FF8888";
	}

	special(map, pacman, ghosts, mapSize)
	{
		this.firstTarget = false;

		if (this.lineOfSight(map, pacman, ghosts, mapSize) && !pacman.getPower())
		{
			this.hasLineOfSight = true;
			for (let ghost in ghosts)
				ghost.setTimer(5 * 60);
		}
		else
			this.hasLineOfSight = false;

		if (!pacman.getPower())
		{
			this.hasTarget = true;
			this.target = pacman.getPos();
		}
	}

	getType()
	{
		return 1;
	}
}

File: C++/SignalGhost.cpp
#include "SignalGhost.h"

SignalGhost::SignalGhost()
{
	color = sf::Color(0xFF8888FF);
}

void SignalGhost::special(std::vector<std::vector<char>>& map, std::vector<Pacman>& pacmen, std::vector<Ghost*>& ghosts, sf::Vector2i mapSize)
{
	hasLineOfSight = false;
	firstTarget = false;

	for (int i = 0; i < pacmen.size(); i++)
	{
		if (lineOfSight(map, pacmen[i], ghosts, mapSize) && !pacmen[i].getPower() && isFree)
		{
			hasLineOfSight = true;
			for (auto& ghost : ghosts)
			{
				ghost->setTimer(5 * 60);
				ghost->assignPacman(i);
			}
		}

		if (!pacmen[targetPacman].getPower())
		{
			hasTarget = true;
			target = pacmen[targetPacman].getPos();
			if (!*pacmen[targetPacman].lives)
				targetPacman = (targetPacman + 1) % pacmen.size();
		}
	}
}

void SignalGhost::assignPacman(int index)
{
	if (firstTarget)
		targetPacman = index;
}
